tck/src/main/java/org/reactivestreams/tck/TestEnvironment.java

```diff
@@ -1025,22 +1072,44 @@ public void expectCompletion(long timeoutMillis, String errorMsg) throws Interru
       } // else, ok
     }
 
-    @SuppressWarnings("unchecked")
+    /**
+     * @deprecated Deprecated in favor of {@link #expectError(Class, long, long, String)}.
+     */
+    @Deprecated
     public <E extends Throwable> E expectError(Class<E> clazz, long timeoutMillis, String errorMsg) throws Exception {
-      Thread.sleep(timeoutMillis);
-
-      if (env.asyncErrors.isEmpty()) {
-        return env.flopAndFail(String.format("%s within %d ms", errorMsg, timeoutMillis));
-      } else {
-        // ok, there was an expected error
-        Throwable thrown = env.asyncErrors.remove(0);
+      return expectError(clazz, timeoutMillis, timeoutMillis, errorMsg);
+    }
 
-        if (clazz.isInstance(thrown)) {
-          return (E) thrown;
+    @SuppressWarnings("unchecked")
+    final <E extends Throwable> E expectError(Class<E> clazz, final long totalTimeoutMillis,
+                                              long pollTimeoutMillis,
+                                              String errorMsg) throws Exception {
+      long totalTimeoutRemainingMillis = totalTimeoutMillis;
+      long timeStampA = System.nanoTime();
+      long timeStampB;
+
+      for (;;) {
+        Thread.sleep(Math.min(pollTimeoutMillis, totalTimeoutRemainingMillis));
```

Comment:
Using ns resolution we'll need to either use https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html#sleep-long-int- or `LockSupport.parkNanos()`

Reply:
The intention was not to sleep with NS resolution. IMHO millis should be sufficient for this use case as it is an approximation anyways (and used for testing).